packages/govern-contract-utils/test/erc1167-proxy-factory.test.ts

```diff
@@ -0,0 +1,120 @@
+  CloneFactoryMockFactory,
+  CloneFactoryMock,
+  ClonedContract,
+  ClonedContractWithInit
+} from '../typechain'
+import
+  * as ClonedContractArtifact
+  from '../artifacts/contracts/test/ClonedContract.sol/ClonedContract.json'
+import
+  * as ClonedContractWithInitArtifact
+  from '../artifacts/contracts/test/ClonedContractWithInit.sol/ClonedContractWithInit.json'
+
+
+describe('ERC1167ProxyFactory', () => {
+  let signers: Signer[],
+    owner: string,
+    factory: CloneFactoryMock,
+    clonedContractAddress: string,
+    clonedContract: any
+
+  before(async () => {
+    signers = await ethers.getSigners()
+    owner = await signers[0].getAddress()
+    const CloneFactoryMock = (await ethers.getContractFactory('CloneFactoryMock')) as CloneFactoryMockFactory
+    factory = await CloneFactoryMock.deploy()
+  })
+
+  context('Clone', () => {
+    it('clone without constructor parameters', async () => {
+      await factory.clone()
+
+      clonedContractAddress = await factory.latestClonedContract()
+      clonedContract = (await ethers.getContractAt(
+        ClonedContractArtifact.abi,
+        clonedContractAddress,
+        signers[0]
+      )) as ClonedContract
+
+      expect(await clonedContract.getRandomString())
+        .to.be.equal('NO INIT')
+    })
+
+    it('clone with constructor parameters', async () => {
+      await factory.cloneWithInitData()
+
+      clonedContractAddress = await factory.latestClonedContract()
+      clonedContract = (await ethers.getContractAt(
+        ClonedContractWithInitArtifact.abi,
+        clonedContractAddress,
+        signers[0]
+      )) as ClonedContractWithInit
+
+      expect(await clonedContract.randomString())
+        .to.be.equal('INIT DATA')
+    })
+  })
+
+  context('Clone with create2', () => {
+    it('clone without constructor parameters', async () => {
+      await factory.clone2()
+
+      clonedContractAddress = await factory.latestClonedContract()
+      clonedContract = (await ethers.getContractAt(
+        ClonedContractArtifact.abi,
+        clonedContractAddress,
+        signers[0]
+      )) as ClonedContract
+
+      expect(await clonedContract.getRandomString())
+        .to.be.equal('NO INIT')
+    })
+
+    it('clone with constructor parameters', async () => {
+      await factory.clone2WithInitData()
+
+      clonedContractAddress = await factory.latestClonedContract()
+      clonedContract = (await ethers.getContractAt(
+        ClonedContractWithInitArtifact.abi,
+        clonedContractAddress,
+        signers[0]
+      )) as ClonedContractWithInit
+
+      expect(await clonedContract.randomString())
+        .to.be.equal('INIT DATA')
+    })
+  })
+
+  context('Helper methods', () => {
+    it('calls "generateCode" as expected', async () => {
+      await factory.generateCode()
+
+      expect(await factory.generatedCode())
+        .to.equal('0x3d602d80600a3d3981f3363d3d373d3d3d363d739467a509da43cb50eb332187602534991be1fea45af43d82803e903d91602b57fd5bf3')
```

Comment:
We get this value from the artifact? Or check manually the generate code?

Reply:
I decided to not implement the ``generateCode`` logic in JS to lower the complexity.. but I'm open to doing it.. The thing is we would just test the solidity language itself in the end.. This means the test helpers I wrote for ``ERC3000Data`` for example are also not really required but I felt like doing them as a short refresher after the seven months sabbatical..

Reply:
> Or check manually the generate code?

... and to answer your question:

I didn't have to check the returned binary of ``generateCode`` itself cause the above test cases do already assert that the generated code does work as expected and ``clone`` does deploy a proxy contract with his own state. This means if ``generateCode`` doesn't work properly do also the test cases above not work because it is already tested indirectly.

Reply:
Great, thanks for the explanation.